gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -398,68 +398,74 @@ OGRFeature *OGRFlatGeobufLayer::GetFeature(GIntBig nFeatureId)
     }
+
     if (VSIFSeekL(m_poFp, sizeof(magicbytes), SEEK_SET) == -1) // skip magic bytes
         return CPLErrorIO();
     uoffset_t headerSize;
     if (VSIFReadL(&headerSize, sizeof(uoffset_t), 1, m_poFp) != 1)
         return CPLErrorIO();
     CPL_LSBPTR32(&headerSize);
-    auto featuresCount = m_poHeader->features_count();
+
     try {
-        auto treeSize = PackedRTree::size(featuresCount);
-        OGREnvelope env;
-        m_poFilterGeom->getEnvelope(&env);
-        Rect r { env.MinX, env.MinY, env.MaxX, env.MaxY };
-        CPLDebug("FlatGeobuf", "Spatial index search on %f,%f,%f,%f", env.MinX, env.MinY, env.MaxX, env.MaxY);
-        auto readNode = [this, headerSize] (uint8_t *buf, size_t i, size_t s) {
-            if (VSIFSeekL(m_poFp, sizeof(magicbytes) + sizeof(uoffset_t) + headerSize + i, SEEK_SET) == -1)
-                throw std::runtime_error("I/O seek failure");
-            if (VSIFReadL(buf, 1, s, m_poFp) != s)
-                throw std::runtime_error("I/O read file");
-        };
-        m_foundFeatureIndices = PackedRTree::streamSearch(featuresCount, indexNodeSize, r, readNode);
-        m_featuresCount = m_foundFeatureIndices.size();
-        if (m_featuresCount == 0) {
-            CPLDebug("FlatGeobuf", "Spatial index query found no features");
-            return OGRERR_NONE;
+        auto treeSize = indexNodeSize > 0 ? PackedRTree::size(featuresCount) : 0;
+        if (treeSize > 0 && m_poFilterGeom && !m_ignoreSpatialFilter) {
+            CPLDebug("FlatGeobuf", "Attempting spatial index query");
+            OGREnvelope env;
+            m_poFilterGeom->getEnvelope(&env);
+            Rect r { env.MinX, env.MinY, env.MaxX, env.MaxY };
+            CPLDebug("FlatGeobuf", "Spatial index search on %f,%f,%f,%f", env.MinX, env.MinY, env.MaxX, env.MaxY);
+            auto readNode = [this, headerSize] (uint8_t *buf, size_t i, size_t s) {
+                if (VSIFSeekL(m_poFp, sizeof(magicbytes) + sizeof(uoffset_t) + headerSize + i, SEEK_SET) == -1)
+                    throw std::runtime_error("I/O seek failure");
+                if (VSIFReadL(buf, 1, s, m_poFp) != s)
+                    throw std::runtime_error("I/O read file");
+            };
+            m_foundFeatureIndices = PackedRTree::streamSearch(featuresCount, indexNodeSize, r, readNode);
+            m_featuresCount = m_foundFeatureIndices.size();
+            CPLDebug("FlatGeobuf", "%lu features found in spatial index search", static_cast<long unsigned int>(m_featuresCount));
+            m_queriedSpatialIndex = true;
+        }
+
+        if (!m_featureOffsets) {
+            CPLDebug("FlatGeobuf", "Seek to feature offsets index position");
+            if (VSIFSeekL(m_poFp, sizeof(magicbytes) + sizeof(uoffset_t) + headerSize + treeSize, SEEK_SET) == -1)
+                return CPLErrorIO();
+            m_featureOffsets = static_cast<uint64_t *>(VSI_MALLOC_VERBOSE(featureOffsetsSize));
+            if (!m_featureOffsets)
```

Comment:
There seem to be mixture of `if (m_poFp == nullptr)` and with implicit `nullptr`.
Although `if (!pointer) {...}` is canonical C++, it may be worth to decide on one way and keep things consistent.